### ISSUE TITLE
Make Concourse compatible with Credhub 2.1

### ISSUE THREAD

**Issue by jama22** [concourse/concourse#2723]
Investigate breaking changes from CredHub 2.1 and make Concourse compatible with CredHub 2.1

Relevant information from CredHub team:
Release Notes: https://github.com/pivotal-cf/credhub-release/releases/tag/2.0.0
API Documentation: http://credhub-api.cfapps.io/version/2.1/

- Read the docs
- Investigate and outline changes to Concourse required to get it working
- Update pipelines
- ???
- Run multi-tenant CredHub on Wings

---

[Update from @cirocosta]:

- [x] Get PR in - https://github.com/cloudfoundry-incubator/credhub-cli/pull/72
- [x] Test integration with both versions of Credhub server in `topgun`

**Comment by jama22** [comment]
Here's additional context from an old conversation circa June 2018

As a part of bumping our major version, we will be making some backwards-incompatible API and manifest changes. The biggest ones are as follows:

- Authorization model changes: Identities will no longer be able to write to CredHub solely by virtue of authenticating. Instead, an explicit access control entry will need to be created, granting an identity read, write, delete, read_acl, and/or write_acl permissions to an identity, for a given namespace.
- `set` and `generate` changes: CredHub is simplifying the rules that determine when `set` or `generate` requests will do nothing or overwrite existing values.
- Minor manifest changes: The structure of our properties hash is changing slightly.
- Authorization on by default: The default value of `credhub.authorization.acls.enabled` is changing from false to true.
- For bosh manifests that deploy and configure CredHub, necessary changes will include:
- Restructuring the properties hash
- Adding explicit permissions for certain users to CredHub's properties hash (similarly to configuring UAA clients)
- CredHub will ensure that the migration from 1.9, 1.7 or 1.6 to 2.x will not result in data loss for users. Old permissions will be migrated such that the platform will continue to function throughout the upgrade.

**Comment by scottietremendous** [comment]
Going to break down the first step into something more specific. We want to prioritize existing compatibility with our supported versions. We'd start with 4.2.3 and validate that this works. After that we'd probably want to check out the 5.x series. 3.13 will be out of scope for this at this moment.

**Comment by cirocosta** [comment]
Hey, I just noticed that there's this other issue related to CredHub 2.x+: https://github.com/concourse/concourse/issues/2293

**Comment by chenbh** [comment]
Findings so far:
- Confirmed Concourse 4.2.3 works with Credhub server up to 1.9.9 (latest 1.9.X), but not 2.0.0
- Credhub CLI, the Go library we are using, has backwards compatibility to server 1.X from CLI 2.0.1 and up (CLI 2.0.0 only compatible with server 2.X)
- We are using a custom commit version between CLI 2.0.0 and 2.0.1 that technically shouldn't even work with 1.X but should be fine for 2.X????
- It should be a simple dependency bump for Concourse 5.X to be compatible with Credhub server 2.X (which should also include backwards compatibility with server 1.X). I have started on this but it is currently paused until I figure out how Bosh works

**Comment by chenbh** [comment]
Update: finally figured out what this whole mess is about

Context:
- Credhub is separated into 2 components, the [server](https://github.com/cloudfoundry-incubator/credhub) and the [CLI](https://github.com/cloudfoundry-incubator/credhub-cli)
- The CLI also contains a Go library to interact with the server directly
- We are using the server in our `topgun` tests and the CLI library in our actual code
- Credhub CLI does not follow `go mod` semantics (which requires all releases to be prefix'ed with `v`) which makes `go mod` to be unable to resolve a proper version and instead use a commit

Issue:
- There are configuration changes between server 1.X and 2.X, this is the reason why `topgun` breaks
- (Credhub) CLI 2.0.0 supports server 2.X but is incompatible with server 1.X
- (Credhub) CLI 2.0.1 is compatible with both server 1.X and 2.X
- Credhub CLI 2.0.1 contains a [commit](https://github.com/cloudfoundry-incubator/credhub-cli/commit/49f312b0923e0fc2afe9ea9711c65533cbd78b94) that made it impossible to differentiate between `credhub.Error`s (cred not found, insufficient permissions) or generic `Error`s (network issues)
- Network issues should cause a failure but a Credhub error is recoverable and acceptable (e.g. if we fail to get the secret at the pipeline level we should still be allowed to look for the secret at the team level)
- We are using some commit between Credhub CLI 2.0.0 and 2.0.1 that apparently supports server 1.X but is still able to differentiate between `credhub.Error` and `Error`

Solution:
- Test integration with both versions of Credhub server in `topgun`
- Ping CredHub for support and maybe submit a PR to fix the generic Error

**Comment by chenbh** [comment]
opened PR in credhub with the fix https://github.com/cloudfoundry-incubator/credhub-cli/pull/72

**Comment by jama22** [comment]
@chenbh would we have the same headaches if we skipped CredHub 2.0 support and went direct to CredHub 2.1?

**Comment by chenbh** [comment]
@jama22 yeah the same issues should occur on any version of 2.x for both server and CLI, I used the specific version numbers to narrow down the details in case anyone was interested

**Comment by scottietremendous** [comment]
Let's fix this into a release instead of the SHA going forward.

**Comment by pivotal-bin-ju** [comment]
**Dev Notes**
- We've understand the issue, and plan has been made for solving this issue. 
   We are going to bump the credhub cli version to the latest (2.4.0) that fixed the generic error issue. After that we're going to run topgun test to verify that concourse can talk to credhub 1.x and credhub 2.x.
- We are trying to understand how to test locally.
- We're currently looking at the `bosh-topgun` job to figure out how to run topgun locally, and potentially use that test suite to deploy concourse and credhub for us onto a bosh deployment.

**Comment by scottietremendous** [comment]
Once this is complete and verified, I think it would be a good idea to create issues around:

- Adding Credhub 2.1 to Wings
- Documenting multi-tenancy with Credhub 2.1 on Concourse

**Comment by YoussB** [comment]
- ran a basic topgun test against 1.9.5, 1.9.12 (latest 1.9.x) and 2.2.0 (latest 2.x)
- 1.9 are passing the test
- there is a configuration problem with 2.2.0 that I am still trying to work around.
```
Error: Unable to render instance groups for deployment. Errors are:
  - Unable to render jobs for instance group 'credhub'. Errors are:
    - Unable to render templates for job 'credhub'. Errors are:
      - Error filling in template 'validation_authorization.yml.erb' (line 8: When ACLs are enabled you must provide at least one permission so that some actor can access CredHub. Please update your manifest to proceed.)
      - Error filling in template 'validation_encryption.yml.erb' (line 62: undefined method `key?' for nil:NilClass)
      - Error filling in template 'application_encryption.yml.erb' (line 27: undefined method `[]' for nil:NilClass)Updating deployment:
  Expected task '1351264' to succeed but state is 'error'
```

**Comment by mhuangpivotal** [comment]
With new configuration syntax mentioned above, the topgun test is passing locally for `credhub` releases 1.9.5, 1.9.12 (latest 1.9.x), and 2.3.0 (latest), running against concourse v5.1.0.
